packages/components/src/progress-bar/test/index.tsx

```diff
@@ -56,7 +56,7 @@ describe( 'ProgressBar', () => {
 		const indicator = container.firstChild?.firstChild;
 
 		expect( indicator ).toHaveStyle( {
-			width: '55%',
+			'--indicator-width': '55%',
```

Comment:
We will be able to revert this if we use the prop directly.

Reply:
Correct.

Reply:
I'd add that I don't think this test should exist, as it's testing implementation details and not necessarily observable behavior. Instead, this should be tested through visual snapshot testing or similar IMO.

Reply:
Hm, I'd disagree this is testing an implementation detail. The [definition](https://kentcdodds.com/blog/testing-implementation-details#so-what-are-implementation-detailsthen) by Kent C. Dodds says:

> Implementation details are things which users of your code will not typically use, see, or even know about.

I'd argue this isn't the case for the width of the progress bar indicator - the indicator is actually defined by its width for the user, and it's definitely the primary thing the user will see from that component.

That being said, I still think a variation of this test is necessary, but it doesn't necessarily have to test a CSS var, but rather, the width of the element for example.

Reply:
The width is resolved by the web rendering engine, which jsdom doesn't have, so it's not possible to test this in the context of a unit test. Leaving the (potentially very interesting!) philosophical debate about what qualifies as an implementation detail, in this specific case I'd highlight the fact that what the test is doing (in the previous and current versions) is test React/emotion for the most part, and whether when you set a value, the value is set. For that alone I don't see much value.

Regarding implementation details (and with no intention to prolong this more than necessary, just for fun :P), I'd argue that what's observable here is what the user sees or whether pixels are rendered a certain way. My understanding of testing observable behavior is that, if you change the implementation, the test doesn't need to be modified. In this case, if we were to switch to an SVG path based implementation, or maybe a CSS transform one, we'd need to change the test since it directly depends on the way it is implemented.

You can definitely test the observable part here though, through visual snapshot testing, or even with a headless browser that does implement a rendering engine (cypress, puppeteer, etc).

Reply:
I think we're more or less on the same page about what an implementation detail is, but it's also about compromise in favor of keeping as few e2e tests as possible, because they come at their own added cost. Is an e2e test more useful in this case, being less about the actual implementation and more about what the user sees - absolutely. Is it more expensive to have and maintain - yes. Thus the compromise. 

On another note, visual regression testing is definitely something we've been aiming to have for some scenarios, and it can be handy as we iterate on components. I know @mirka has been experimenting with that before.

Reply:
While testing styles in Jest are indeed a lower-cost compromise over a full vizreg test, it also isn't cost-free in the sense that it can add a maintenance cost that can be arguably disproportionate to the low confidence that it adds — similar to a snapshot test.

In this particular case I don't think it's likely that someone would break the style and land that PR without anyone noticing (at least for the happy path), so to be honest I don't care that much about whether or how we do the test. If it does ever break, we can perhaps alter the strategy based on how it broke.